flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -1054,10 +1054,10 @@ protected abstract void internalDeregisterApplication(
 	 * Allocates a resource using the resource profile.
 	 *
 	 * @param resourceProfile The resource description
-	 * @return Collection of {@link ResourceProfile} describing the launched slots
+	 * @return whether the resource can be allocated
 	 */
 	@VisibleForTesting
-	public abstract Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile);
+	public abstract boolean startNewWorker(ResourceProfile resourceProfile);
```

Comment:
What happens if the RM can only start a multiple of `resourceProfile`? Wouldn't we calculate the number of available slots wrongly in this case?

Reply:
In this commit, we have moved the logics of calculating number of new pending slots into `SlotManagerImpl`. Therefore, `SlotManagerImpl` only needs to know whether the new worker is started or not, and should be able to correctly calculate the number of available slots.

In the final state, RM accepts `WorkerResourceSpec` instead of slot `ResourceProfile` for starting new workers. It would become a SM decision that how many multiple of `ResourceProfile` the worker should be start with.

Reply:
Ok if I understand it correctly, then this PR relies on the fact that the `RM` and the `SlotManager` both get the same (implicitly or explicitly) `WorkerResourceSpec` assigned, right?

I think it can still happen that the RM won't be able to start exactly a container of the size `WorkerResourceSpec`, even after later changes. Imagine that we are on Yarn were the requested `WorkerResourceSpec` is below the minimum. How should Flink behave in this situation? Should we fail? Should we leave some of the resources unused? Or should we tell the `SlotManager` about it so that it can take the additional resources into account?

Reply:
True, it is possible that RM won't be able to start a container for given `WorkerResourceSpec`.

I think if the requested resource, to be specific the `TaskExecutorProcessSpec` derived from `WorkerResourceSpec`, is smaller than the Yarn min-allocation, we can simply throw a warning for that. Same for when requested resource is not an integer multiple of min-allocation. The critical part is if the requested resource is larger than the max-allocation, and we should check return false in such cases.

I think these checks should be performed by the RM implementations. I'll add the checks for Yarn in #11353. For K8s, I think there are no such min/max-allocation limits, but I'll double-check on that.

Reply:
Hmm, but with this API we will waste resources because the caller thinks it has started a worker of the specified size, right? An idea how to improve the situation is to let the `startNewWorker` method return the size of the container it has started. This would allow the caller to react to it.

Reply:
Looking more into this, I'm not sure whether returning the container size is the right approach.

The problem is, the container resource contains two parts that are decided by different components.
- Those contained in `WorkerResourceSpec` are decided by `SlotManager`: cpu, task heap, task off-heap, network, managed.
- Those contained in `TaskExecutorProcessSpec` but not in `WorkerResourceSpec` are decided by `ResourceManager`: framework heap, framework off-heap, jvm metaspace, jvm overhead.

If we return the size of the entire started container to SM, it does not help SM to decide whether there are resources wasted, because SM is not aware of how many resources in addition to `WorkerResourceSpec` that the framework and jvm need.

An alternative is to exclude the framework and jvm resources and the remaining to SM. The problem is that, the excluded part is not constant and might be dependent on `WorkerResourceSpec`. To be specific, jvm overhead might be configured as a fraction of total process memory size. That means increasing the total size of `WorkerResourceSpec` may also lead to the increasing of jvm overhead.

I think what should be returned by this API really depends on how we want to solve the resource wasting issue. TBH, I'm not sure when would we have a `SlotManager` implementation that deals with the Yarn resource wasting problem. I would suggest to create a follow-up ticket and leave it to when we do have such a SM implementation. WDYT?

Reply:
Sounds good to me. Thanks for the background information.